File: companion/github-yayin/bolumler/b01/ornek-01/cozum.py
import argparse
import math
import platform

import pandas as pd


def hesapla(veri):
    if list(veri.columns) != ["devam_saati", "basari", "program"]:
        raise ValueError("CSV sutunlari ve sirasi veri sozluguyle uyusmuyor.")
    if veri.empty or veri.isna().any().any():
        raise ValueError("Bu ornek bos veri veya eksik hucre kabul etmiyor.")
    if not veri["program"].isin(["A", "B"]).all():
        raise ValueError("Program etiketleri A veya B olmali.")
    veri = veri.copy()
    for alan in ["devam_saati", "basari"]:
        veri[alan] = pd.to_numeric(veri[alan], errors="raise")
        if not veri[alan].map(math.isfinite).all():
            raise ValueError("Sayisal degerler sonlu olmali.")
    if (veri["devam_saati"] < 0).any():
        raise ValueError("Devam suresi negatif olamaz.")
    veri["program"] = pd.Categorical(veri["program"], categories=["A", "B"])
    satirlar = [
        ("veri", "satir_sayisi", len(veri)),
        ("veri", "degisken_sayisi", len(veri.columns)),
    ]
    for alan in ["devam_saati", "basari"]:
        degerler = veri[alan]
        satirlar.extend([
            (alan, "gecerli", degerler.count()),
            (alan, "eksik", degerler.isna().sum()),
            (alan, "ortalama", degerler.mean()),
            (alan, "en_kucuk", degerler.min()),
            (alan, "en_buyuk", degerler.max()),
        ])
    satirlar.extend([
        ("program", "gecerli", veri["program"].count()),
        ("program", "eksik", veri["program"].isna().sum()),
    ])
    for program in ["A", "B"]:
        frekans = (veri["program"] == program).sum()
        satirlar.extend([
            ("program", program + "_frekans", frekans),
            ("program", program + "_oran", frekans / len(veri)),
        ])
    return pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])


def main():
    parser = argparse.ArgumentParser(description="B01 ortak CSV cozumu")
    parser.add_argument("--check", action="store_true",
                        help="Sonuclari beklenen degerlerle karsilastir")
    args = parser.parse_args()
    veri = pd.read_csv("veri.csv", dtype={"program": "string"})
    sonuc = hesapla(veri)
    print(f"Python {platform.python_version()} | pandas {pd.__version__}")
    print(sonuc.to_string(index=False))
    if args.check:
        beklenen = pd.read_csv("beklenen-sonuclar.csv")
        anahtarlar = ["degisken", "olcu"]
        if not sonuc[anahtarlar].equals(beklenen[anahtarlar]):
            raise ValueError("Beklenen sonuc satirlari veya sirasi farkli.")
        for hesaplanan, hedef in zip(sonuc["deger"], beklenen["deger"]):
            if not math.isclose(hesaplanan, hedef, rel_tol=1e-9, abs_tol=1e-9):
                raise ValueError(f"Sonuc uyusmazligi: {hesaplanan} != {hedef}")
        print(f"DOGRULANDI: {len(beklenen)} kontrol degeri eslesiyor.")


if __name__ == "__main__":
    main()

File: companion/bolumler/b01/ornek-01/cozum.R
veri <- read.csv("veri.csv", stringsAsFactors = FALSE,
                 fileEncoding = "UTF-8")
stopifnot(identical(names(veri),
                    c("devam_saati", "basari", "program")),
          nrow(veri) > 0, !anyNA(veri),
          all(veri$program %in% c("A", "B")),
          is.numeric(veri$devam_saati), is.numeric(veri$basari),
          all(is.finite(veri$devam_saati)),
          all(is.finite(veri$basari)),
          all(veri$devam_saati >= 0))
veri$program <- factor(veri$program, levels = c("A", "B"))
sonuc <- data.frame(
  degisken = "veri",
  olcu = c("satir_sayisi", "degisken_sayisi"),
  deger = c(nrow(veri), ncol(veri))
)
for (alan in c("devam_saati", "basari")) {
  degerler <- veri[[alan]]
  ozet <- data.frame(
    degisken = alan,
    olcu = c("gecerli", "eksik", "ortalama", "en_kucuk", "en_buyuk"),
    deger = c(sum(!is.na(degerler)), sum(is.na(degerler)),
              mean(degerler), min(degerler), max(degerler))
  )
  sonuc <- rbind(sonuc, ozet)
}
sonuc <- rbind(sonuc, data.frame(
  degisken = "program", olcu = c("gecerli", "eksik"),
  deger = c(sum(!is.na(veri$program)), sum(is.na(veri$program)))
))
for (program in c("A", "B")) {
  frekans <- sum(veri$program == program)
  sonuc <- rbind(sonuc, data.frame(
    degisken = "program",
    olcu = paste0(program, c("_frekans", "_oran")),
    deger = c(frekans, frekans / nrow(veri))
  ))
}
cat(R.version.string, "\n")
print(sonuc, row.names = FALSE)
if ("--check" %in% commandArgs(trailingOnly = TRUE)) {
  beklenen <- read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE)
  stopifnot(nrow(sonuc) == nrow(beklenen),
            all(sonuc$degisken == beklenen$degisken),
            all(sonuc$olcu == beklenen$olcu),
            isTRUE(all.equal(sonuc$deger, beklenen$deger,
                             tolerance = 1e-9, check.attributes = FALSE)))
  cat("DOGRULANDI:", nrow(beklenen), "kontrol degeri eslesiyor.\n")
}